firedrake/slate/slate.py

```diff
@@ -612,3 +611,39 @@ def __repr__(self):
         tensor, = self.operands
         coefficient = self.actee
         return "Action(%r, %r)" % (tensor, coefficient)
+
+    @cached_property
+    def _key(self):
+        """Returns a key for hash and equality."""
+        return (type(self), self.operands, self.actee)
+
+
+def merge_coefficients(operands, assembled_coeffs=None):
+    """Returns a set of non-repeated coefficient objects.
+
+    :arg operands: a `tuple` of Slate tensors.
+    :arg assembled_coeffs: a `tuple` of assembled data provided
+                           as a `firedrake.Function`.
+    """
+    coefficient_list = []
+    extra_data = assembled_coeffs or ()
+    for op in operands:
+        coefficient_list.extend(op.coefficients())
+
+    return tuple(OrderedDict.fromkeys(tuple(coefficient_list) +
+                                      extra_data))
```

Comment:
I think you can do this with less code.

`assembled_coeffs=None` version:
```python
return tuple(OrderedDict.fromkeys(chain(*[op.coefficients() for op in operands])))
```

When `assembled_coefficient` is specified:
```python
return tuple(OrderedDict.fromkeys(chain(*([op.coefficients() for op in operands] + [assembled_coefficient]))))
```
or if you don't want such a long line:
```python
coefficientz = [op.coefficients() for op in operands] + [assembled_coefficient]
return tuple(OrderedDict.fromkeys(chain(*coefficientz)))
```